### PR TITLE
Bump gateway api dep to v0.3.0

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Based on https://github.com/projectcontour/contour-operator/pull/311 and a draft until that is merged

Using commit that matches contour controller

### CODE REVIEW

File: internal/objects/gateway/gateway.go [hunk 1]
```diff
@@ -78,7 +79,7 @@ func ContourForGateway(ctx context.Context, cli client.Client, gw *gatewayv1alph
 		return nil, nil
 	}
 
-	cntr, err := objcontour.CurrentContour(ctx, cli, gc.Spec.ParametersRef.Namespace, gc.Spec.ParametersRef.Name)
+	cntr, err := objcontour.CurrentContour(ctx, cli, pointer.StringPtrDerefOr(gc.Spec.ParametersRef.Namespace, ""), gc.Spec.ParametersRef.Name)
```
Comment (sunjayBhatia) [inline internal/objects/gateway/gateway.go:82]: these were just quick changes based on the switch to pointers, so there might be some cleanup actually needed and new tests added

Comment (skriss) [reply internal/objects/gateway/gateway.go:82]: Should probably be an error if there's no namespace in the `ParametersRef`, given that a `Contour` is a namespaced CRD.

**Review by skriss — CHANGES_REQUESTED** [review]
Ready to bump to v0.3.0 now! One small comment otherwise this looks good to me.

File: internal/objects/gateway/gateway.go [hunk 2]
```diff
@@ -78,7 +78,15 @@ func ContourForGateway(ctx context.Context, cli client.Client, gw *gatewayv1alph
 		return nil, nil
 	}
 
-	cntr, err := objcontour.CurrentContour(ctx, cli, gc.Spec.ParametersRef.Namespace, gc.Spec.ParametersRef.Name)
+	if gc.Spec.ParametersRef == nil {
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/objects/gateway/gateway.go:81]: seems reasonable to return an error to me

**Review by skriss — APPROVED** [review]
This all looks reasonable to me. The upgrade e2e failed, re-running, it's likely a flake.

File: internal/objects/gateway/gateway.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danehans) [reply internal/objects/gateway/gateway.go:81]: `ParametersRef `is [optional](https://gateway-api.sigs.k8s.io/references/spec/#networking.x-k8s.io/v1alpha1.GatewayClassSpec) and a gc will pass validation accordingly. However, the operator panics for this use case. I created https://github.com/projectcontour/contour-operator/issues/337 to capture this issue.

**Review by stevesloka — APPROVED** [review]

**Pull request merged**